### PR TITLE
[`Flex Attn`] Fix torch 2.5.1 incompatibilities

### PR DESCRIPTION
**Pull request by vasqu**
Flex attention with `torch==2.5.1` is broken, see e.g. https://github.com/huggingface/transformers/issues/37380#issuecomment-2790650478

The PR has 3 changes:
1. Removes the compilation on the block mask as it does not provide any major benefits in perf either way and seems to be the root cause behind the issues (torch 2.5.x specific). Open to making it version dependable but thought it's easier this way for now.
2. Change the mask padding to be not addition but up to a value, e.g. 4 -> 4 + 6 -> 10 (before) vs 4 -> 6 (after); unsure if this is breaking but seemed appropriate. Edit: changed this back to old behavior.
3. Mask padding goes to at least the value of the default block size in flex attention (can be observed on smaller seq_lens leading to indexing errors) (torch 2.5.x specific).

# What does this PR do?

Fixes # (issue)

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [x] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/main/CONTRIBUTING.md#create-a-pull-request),
      Pull Request section?
- [ ] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to it if that's the case.
- [ ] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/transformers/tree/main/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/main/docs#writing-source-documentation).
- [ ] Did you write any new necessary tests?

## Who can review?

Anyone in the community is free to review the PR once the tests have passed. Feel free to tag
members/contributors who may be interested in your PR.

cc @ArthurZucker @Cyrilvallez

### CODE REVIEW

File: src/transformers/integrations/flex_attention.py [hunk 1]
```diff
@@ -71,7 +72,7 @@ def __init__(self, training):
                     flex_attention, dynamic=False, mode="max-autotune-no-cudagraphs"
                 )
             else:
-                self._compiled_flex_attention = torch.compile(flex_attention)
+                self._compiled_flex_attention = torch.compile(flex_attention, dynamic=False)
```
Comment (vasqu) [inline src/transformers/integrations/flex_attention.py:75]: Seems to be necessary and can avoid symbolic shape errors to some degree. Decoding/Inference seems to hit issues either way (with this it's recompilations) which are (better) solved in torch 2.6.x.

File: src/transformers/integrations/flex_attention.py [hunk 2]
```diff
@@ -115,7 +116,9 @@ def make_flex_block_causal_mask(
         key_length = total_seq_len
     if not query_length:
         query_length = total_seq_len
-    attention_mask_2d = torch.nn.functional.pad(attention_mask_2d, value=0, pad=(0, key_length))
+    attention_mask_2d = torch.nn.functional.pad(
+        attention_mask_2d, value=0, pad=(0, abs(total_seq_len - max(key_length, flex_default_block_size)))
```
Comment (winglian) [inline src/transformers/integrations/flex_attention.py:120]: should this be `abs(...)` or the `max(0, ...)`?

Comment (vasqu) [reply src/transformers/integrations/flex_attention.py:120]: Not exactly sure what you mean (the second term?); could you expand the formula you mean? I expect that the passed kv len is at least the size of the attn mask (seq len).

The intention behind this current form is that it pads to a minimum of `default_block_size` (otherwise flex attn creates an internally to big block leading to index issues) or the passed kv_len if necessary (2nd term in `pad=`). Q doesn't seem to need padding (1st term in `pad=`).

File: src/transformers/integrations/flex_attention.py (hunk 1 above)
Comment (Cyrilvallez) [reply src/transformers/integrations/flex_attention.py]: Are we sure this does not degrade performance on recent torch versions and models? 🤔 May need to be guarded against version

**Review by Cyrilvallez — COMMENTED** [review]
Hey @vasqu! Are you sure that `dynamic=False` has no impact on performance in general? That seems unlikely to me 🤔
Also, `make_flex_block_causal_mask` is deprecated in favor of the new mask API, but happy to fix it while we patch the older models still using it!

File: src/transformers/integrations/flex_attention.py (hunk 1 above)
Comment (vasqu) [reply src/transformers/integrations/flex_attention.py]: Hmm, tbh unsure haven't checked. I can add a version check for sure, the PR is already a bit older so I haven't had the new releases (2.7.x+) in mind :D

Comment (vasqu) [reply src/transformers/integrations/flex_attention.py]: Added proper version checks!

**Comment by vasqu** [comment]
@Cyrilvallez since the new mask API deals with similar issues, I've added similar fixes to it 

Tested this on torch versions: `2.5.1, 2.6.0, and 2.7.0`

File: src/transformers/masking_utils.py [hunk 3]
```diff
@@ -550,6 +551,11 @@ def flex_attention_mask(
 
     # Potentially add the padding 2D mask
     if attention_mask is not None:
+        # Older torch (2.5.x) cannot handle sequences < 128 (default block size)
+        # Hence we pad with this as a minimum to ensure this
+        if not _is_torch_greater_or_equal_than_2_6:
+            attention_mask = torch.nn.functional.pad(attention_mask, value=0, pad=(0, flex_default_block_size))
+
```
Comment (Cyrilvallez) [inline src/transformers/masking_utils.py]: Here we always pad `block_size` more, independently of the current size - it works in general as we do not really care if the mask is longer for flex, but just wanna make sure it was intentional not to check current size!

Comment (vasqu) [reply src/transformers/masking_utils.py]: Yea, it was intentional since it won't make much of a difference imo. Added a check on the padding length based on the block size, lmk if that's better

**Review by Cyrilvallez — APPROVED** [review]
Alright, LGTM! Thanks! 🤗

File: src/transformers/masking_utils.py [hunk 4]
```diff
@@ -551,9 +551,10 @@ def flex_attention_mask(
 
     # Potentially add the padding 2D mask
     if attention_mask is not None:
-        # Older torch (2.5.x) cannot handle sequences < 128 (default block size)
-        # Hence we pad to this as a minimum to ensure this
-        pad_len = max(0, flex_default_block_size - attention_mask.shape[1])
+        # Older torch (2.5.x) cannot handle sequences not in multiples of 128 (default block size)
+        # Hence we pad to multiples of this as a minimum to ensure this
+        pad_len = ((attention_mask.shape[1] // flex_default_block_size) + 1) * flex_default_block_size
+        pad_len = max(0, pad_len - attention_mask.shape[1])
```
Comment (Cyrilvallez) [inline src/transformers/masking_utils.py]: very nit, but I don't think we need this `max` op here (should always be larger or equal)

**Review by Cyrilvallez — APPROVED** [review]
Still good, thanks for double-checking!

File: src/transformers/masking_utils.py (hunk 4 above)
Comment (vasqu) [reply src/transformers/masking_utils.py]: Yup, true let me remove it
